From 64bf0a91f7d05798839eb27f12a1d9cc23e88d1d Mon Sep 17 00:00:00 2001
From: "Joshua M. Clulow" <jmc@joyent.com>
Date: Wed, 25 Jan 2017 23:51:55 +0000
Subject: [PATCH] joyent/manta-hk#3 attempt to sort shards numerically Reviewed
 by: David Pacheco <dap@joyent.com> Approved by: Joshua M. Clulow
 <jmc@joyent.com>

---
 bin/manta-hk | 42 +++++++++++++++++++++++++++++++++++++++++-
 1 file changed, 41 insertions(+), 1 deletion(-)

diff --git a/bin/manta-hk b/bin/manta-hk
index 7e1886d..34e9458 100755
--- a/bin/manta-hk
+++ b/bin/manta-hk
@@ -7,7 +7,7 @@
  */
 
 /*
- * Copyright (c) 2015, Joyent, Inc.
+ * Copyright 2017 Joyent, Inc.
  */
 
 /*
@@ -156,10 +156,50 @@ MantaHk.prototype.do_dumps.help = [
 ].join('\n');
 
 
+function shardSorter(a, b)
+{
+	var a_m, b_m;
+	var a_n, b_n;
+
+	/*
+	 * Split the string into two components: any contiguous leading digits
+	 * are treated as a single integer, and everything after that is
+	 * treated as a regular string.
+	 */
+	a_m = a.match(/^([0-9]*)(.*)$/);
+	b_m = b.match(/^([0-9]*)(.*)$/);
+
+	/*
+	 * Sort first by the second component; i.e., by the part _after_ the
+	 * initial integer ID.  This will group entries with the same domain
+	 * name together in the list.
+	 */
+	if (a_m[2] < b_m[2]) {
+		return (-1);
+	} else if (a_m[2] > b_m[2]) {
+		return (1);
+	}
+
+	/*
+	 * If the domain name is equal, sort the initial integer component
+	 * in ascending order.
+	 */
+	a_n = parseInt(a_m[1], 10);
+	b_n = parseInt(b_m[1], 10);
+	if (a_n < b_n) {
+		return (-1);
+	} else if (a_n > b_n) {
+		return (1);
+	}
+
+	return (0);
+}
+
 function printDumpsByDate(shards, dumps)
 {
 	var dates, shardcols;
 
+	shards.sort(shardSorter);
 	dates = Object.keys(dumps).sort();
 	shardcols = shards[0].length + 1;
 	dates.forEach(function (timestamp) {
-- 
2.21.0

